### PR TITLE
Take data providers into account when counting tests to be executed

### PR DESCRIPTION
**Pull request by losipiuk**
Previously we counted method with data providers as single test
execution. Therefore all methods count was off and logs from
ProgressLoggingListener looked weird in such case.

Now we execute data provider within RequirementsExpanderInterceptor
which does the counting - and increment the seenMethodsCount by number
of paremeters sets.

### CODE REVIEW

**Comment by losipiuk** [comment]
@arhimondr, @sopel39, @kokosing Take a look at this in some free time.

File: tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java [hunk 1]
```diff
@@ -80,6 +88,49 @@ public static int getMethodsCountFromContext(ITestContext context)
...
+            Optional<Object[][]> parametersForMethod = getParametersForMethod(newExpandedMethod);
+            if (parametersForMethod.isPresent()) {
+                seenMethodsCount += parametersForMethod.get().length;
+            } else {
+                seenMethodsCount++;
+            }
+        }
+    }
+
+    private Optional<Object[][]> getParametersForMethod(IMethodInstance method)
```
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java]: Maybe put code responsible for data providers (this reflection code) into separate class and add proper testing.

File: tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java [hunk 2]
```diff
@@ -80,6 +88,49 @@ public static int getMethodsCountFromContext(ITestContext context)
...
+        if (dataProviderClass == null || dataProviderClass == Object.class) {
+            dataProviderClass = method.getMethod().getRealClass();
+        }
+        String dataProviderName = testAnnotation.dataProvider();
+        if (dataProviderName.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<Method> dataProviderMethod = asList(dataProviderClass.getMethods()).stream().filter(
+                m -> {
```
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java]: Do you need to scan all the methods? I thought DataProvider is an interface, where only one method returns an array of parameters.

File: tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java [hunk 3]
```diff
@@ -80,6 +88,49 @@ public static int getMethodsCountFromContext(ITestContext context)
...
+                    DataProvider annotation = m.getAnnotation(DataProvider.class);
+                    return annotation != null && annotation.name().equals(dataProviderName);
+                }
+        ).findFirst();
+        if (dataProviderMethod.isPresent()) {
+            try {
+                return Optional.of((Object[][]) dataProviderMethod.get().invoke(method.getInstance()));
```
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java]: ah... it is a static method, so what if there are multiple data provider static methods, will you just find first of them?

**Comment by kokosing** [comment]
I think that using reflection is errorprone, so unit testing for that is needed. Have you check testng code, maybe they have already something to retrieve test parameters for test?

**Comment by losipiuk** [comment]
@kokosing I looked into testNG code. They have sth but it is very convoluted. Have more responsibilities than just listing parameterSets based on data providers - and requires passing some weird dependencies I do not have here.

File: tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java (hunk 2 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java]: DataProvider is an annotation you can add to the method. It has a name and then data provider can be referenced in test method using \@Test(dataProvider = "dataProviderName").
So we have to scan check all methods with \@DataProvider annotation and pick the one with name matching one used by specific test method.

File: tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java (hunk 3 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/initialization/RequirementsExpanderInterceptor.java]: There should be just one

**Comment by losipiuk** [comment]
is that ok now?

File: tempto-core/src/main/java/com/teradata/tempto/internal/DataProviders.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ */
+public class DataProviders
+{
+
+    /**
+     * Obtains list of parameter sets for test method based on defined data provider.
+     * If no data provider is defined then {@link Optional#empty()} is returned.
+     * This method takes into consideration only data providers defined through annotation.
+     */
+    public static Optional<Object[][]> getParametersForMethod(ITestNGMethod method)
```
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/DataProviders.java]: static?

**Comment by kokosing** [comment]
looks good

**Comment by losipiuk** [comment]
merged

**Pull request closed without merging**